### PR TITLE
[BUG] Shares in non-root are not displayed

### PR DESCRIPTION
**Pull request by parneet-guraya**
Fixes: #4432 

Demo -->

https://github.com/owncloud/android/assets/111801812/623c9c3d-32bd-4465-bb45-3669f8d28af9

The reason for this issue was whenever we create/edit a link. After `POST` request we make a `GET` call to retrieve the share info because it was mentioned [here](https://github.com/owncloud/android/blob/bba83debbb09160065a39fdf67961bd334885687/owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/CreateRemoteShareOperation.kt#L131) that the POST request does not return the response that we need.

But the `GET` request is returning wrong path i.e for a file that exists at `/Folder/file` it returns the path `/file` . So, it always returns a path that a file exists in the root even though it is inside some folder. 

Screenshot from 2024-07-10 13-29-27

Above screenshot shows two responses. First one is the `POST` which has path /Fol/file and the second one is `GET` response which has path /file (which is incorrect)

Screenshot from 2024-07-10 13-30-00

**Also the incorrect response from `GET` lead to duplicate entries in DB -->**
- First entry is stored from the response with path as /file(incorrect)
- Second entry is stored when the `ShareView` is refreshed and upon sync (from remote), instead of getting replaced the new entry gets added because there's path ##difference.

**Solution -->**

Turns out `POST` request does return same response as `GET` does and also with correct `filePath`. So, now the need for `GET` doesn't exist.

### CODE REVIEW

**Comment by parneet-guraya** [comment]
@jesmrec does it also fix #3016?

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/GetRemoteShareOperation.kt
Comment (Aitorbp) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/GetRemoteShareOperation.kt:1]: Why did you delete this class? What was its function and why is it not necessary? @parneet-guraya

Comment (parneet-guraya) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/GetRemoteShareOperation.kt:1]: Initially we do `POST` request to create/update link. Then a `GET` request to get the additional info. The reason for making `GET` request was because (mentioned in code's [comment](https://github.com/owncloud/android/blob/bba83debbb09160065a39fdf67961bd334885687/owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/CreateRemoteShareOperation.kt#L131)) the `POST` request does not return the info that we need. 

There are two reasons to remove this file:

- Turns out `POST` request does return all the info we need same as `GET`. So, no reason to call another `GET` request. Hence, removed the file.

- Also `GET` request returns the `filePath` as if the file exists in the root. Meaning even if a file exist inside a folder, it still would show as if it's in the root.

**Comment by jesmrec** [comment]
> @jesmrec does it also fix https://github.com/owncloud/android/issues/3016?

i don't think so, are different things. The current one affects the list in the shares view, and the #3016 affects the list of files. Anyway, i will take a look in QA phase.

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/GetRemoteShareOperation.kt
Comment (parneet-guraya) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/GetRemoteShareOperation.kt:1]: Okay, I wondered the same myself as I said my only clue is that the second call is for purpose of getting additional info that according to the comment wasn't provided by the first request we do. I tried looking back in the history it was a part of very huge change so couldn't track it back further. 

But, as you can see in my PR's comment we're getting the same response as the second request does, in fact the second one returns the wrong path. So, to answer your question I think it could be necessary to make a second request in the past that the `POST` req doesn't give us the info but at sometime later it was added into the API to return the info too? 

Let me know if this clarify your doubt or you have any questions.

Comment (Aitorbp) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/GetRemoteShareOperation.kt:1]: Ok, I think the same like you

**Review by Aitorbp — COMMENTED** [review]
For my part the work is fine. Only one thing is missing, could you add the calens file? Here you have some information to do it: https://github.com/owncloud/android/blob/master/changelog/README.md

**Comment by parneet-guraya** [comment]
@Aitorbp , Added the changelog :+1:

File: changelog/unreleased/4435 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Bugfix: Fix the Shares screen not updating.
```
Comment (Aitorbp) [inline changelog/unreleased/4435:1]: I would change the title: `Shares in non-root are updated correctly`
In the title we don't need the final point.

File: changelog/unreleased/4435 [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+Fixed the Shares screen not updating instantly when create/edit a link.
```
Comment (Aitorbp) [inline changelog/unreleased/4435:3]: And here I would write: 
`The items of the "Share" view are updated instantly when create/edit a link or share with users or groups in a non-root file.`

File: changelog/unreleased/4435 [hunk 3]
```diff
@@ -0,0 +1,6 @@
+Bugfix: Shares in non-root are updated correctly.
```
Comment (Aitorbp) [inline changelog/unreleased/4435:1]: Please, remove the final point `.`

**Review by Aitorbp — CHANGES_REQUESTED** [review]

**Review by Aitorbp — APPROVED** [review]
@parneet-guraya very good job! I will move to QA!

**Comment by jesmrec** [comment]
Let's QA this

**Comment by jesmrec** [comment]
@parneet-guraya first, thanks for the contribution.

I noticed that the fix applies to the public links, but for sharing with users and groups (also reproducible and reported in the issue as a note). Could you take a look?

**Comment by parneet-guraya** [comment]
> I noticed that the fix applies to the public links, but for sharing with users and groups (also reproducible and reported in the issue as a note). Could you take a look?

@jesmrec , you're right. However, I'm not able to add any shares, it seems the demo account does not have privileges? 

Screenshot from 2024-07-24 02-47-24

This is what I get as a response (even on master) but I'm able to do the same from web just fine. Could you give it a look?

**Comment by jesmrec** [comment]
@parneet-guraya this should not be a matter of permissions because web allows to share. Something is not correct in the app there. I will try to check.

**Comment by jesmrec** [comment]
@parneet-guraya which server instance are you using to check it?

**Comment by jesmrec** [comment]
I was able to reproduce the problem with `ocis.ocis.master.owncloud.works` but not with the oCIS stable 5.0.6. I created an issue in oCIS repo: https://github.com/owncloud/ocis/issues/9681 because the problem is also reproducible with iOS client. Web client works fine. In order not to block the current issue, i'd recommend to use an instance with stable 5.0.6

**Comment by parneet-guraya** [comment]
Hi @jesmrec, I'm using demo instance at `demo.owncloud.com` this could be running on latest master? 

So, to run different version i.e 5.0.6, I guess I have to run it locally?

**Comment by jesmrec** [comment]
You can use this instance with the stable version: `ocis.ocis.production.owncloud.works` (i just found it). With same usernames and pwds.

**Comment by parneet-guraya** [comment]
@jesmrec, Would you mind directing me to the resource to how to do it?

**Comment by jesmrec** [comment]
> @jesmrec, Would you mind directing me to the resource to how to do it?

Sorry, what do you mean? you can access [https://ocis.ocis.production.owncloud.works](https://ocis.ocis.production.owncloud.works) just with the URL, as any other instance. Is that right? if not, let me know.

**Comment by parneet-guraya** [comment]
Okay, I get it now. Thanks :+1:

**Comment by jesmrec** [comment]
@parneet-guraya were you able to finally reproduce the "share with users" problem? let us know if you need some extra help or clarification. thanks!!

**Comment by parneet-guraya** [comment]
> @parneet-guraya were you able to finally reproduce the "share with users" problem? let us know if you need some extra help or clarification. thanks!!

Hi, turns out the `POST` request that we make to share with users and groups returns the wrong path (as if the file always is in root folder). This is same as we create public shares where first a `POST` followed by a `GET` request was made but in that case the second request was returning the wrong path so I removed the request because first request was enough according to the data we need.

But, now in this case the only request that we make to share, returns the wrong path. So, it is a backend issue. I read from the issue created by @jesmrec [comment](https://github.com/owncloud/ocis/issues/9681#issuecomment-2247831945) that mobile clients are using deprecated API so could it be the bug exists in the deprecated api?

A quick fix could be just to make a refresh call from remote after doing the `POST` request for adding the share. Which triggers when a dialog appears for adding new shares and that is the reason we start seeing the items that were not showing earlier. 

Do let me know what you think? @JuancaG05 @Aitorbp

**Comment by jesmrec** [comment]
> But, now in this case the only request that we make to share, returns the wrong path. So, it is a backend issue. I read from the issue created by @jesmrec https://github.com/owncloud/ocis/issues/9681#issuecomment-2247831945 that mobile clients are using deprecated API so could it be the bug exists in the deprecated api?

Don't worry about this, ocs is not deprecrated yet. Problem in the oCIS issue finally comes from other side.

**Comment by Aitorbp** [comment]
> > @parneet-guraya were you able to finally reproduce the "share with users" problem? let us know if you need some extra help or clarification. thanks!!
> 
> Hi, turns out the `POST` request that we make to share with users and groups returns the wrong path (as if the file always is in root folder). This is same as we create public shares where first a `POST` followed by a `GET` request was made but in that case the second request was returning the wrong path so I removed the request because first request was enough according to the data we need.
> 
> But, now in this case the only request that we make to share, returns the wrong path. So, it is a backend issue. I read from the issue created by @jesmrec [comment](https://github.com/owncloud/ocis/issues/9681#issuecomment-2247831945) that mobile clients are using deprecated API so could it be the bug exists in the deprecated api?
> 
> A quick fix could be just to make a refresh call from remote after doing the `POST` request for adding the share. Which triggers when a dialog appears for adding new shares and that is the reason we start seeing the items that were not showing earlier.
> 
> Do let me know what you think? @JuancaG05 @Aitorbp

Hello @parneet-guraya , very good research. As you know, we are using a deprecated API and we don't know if someone could change something in this API. So maybe the fast way at the moment is try to do to refresh as you comment. So, yes, go for it 👍

**Comment by parneet-guraya** [comment]
Hi, pushed the fix :+1:

File: owncloudApp/src/main/java/com/owncloud/android/presentation/sharing/ShareActivity.kt [hunk 4]
```diff
@@ -154,7 +154,10 @@ class ShareActivity : FileActivity(), ShareFragmentListener {
                     is UIResult.Loading -> {
                         showLoadingDialog(R.string.common_loading)
                     }
-                    is UIResult.Success -> {}
+                    is UIResult.Success -> {
+                        // needs to trigger refresh after creation because creation request returns wrong path [issue: https://github.com/owncloud/ocis/issues/9746] resulting list not getting updated.
```
Comment (Aitorbp) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/sharing/ShareActivity.kt:158]: Please could you remove the link in the comment. We will link in the pull request. Also the first word of the comment should be in capital letter. Fix this and ready to go 🥇

**Review by Aitorbp — CHANGES_REQUESTED** [review]
@parneet-guraya a few changes more and ready to go 💯

File: owncloudApp/src/main/java/com/owncloud/android/presentation/sharing/ShareActivity.kt (hunk 4 above)
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/sharing/ShareActivity.kt:158]: And please, fix indentation.

Comment (parneet-guraya) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/sharing/ShareActivity.kt]: @jesmrec I'm using reformat code (Ctrl+Alt+L) , is there any other I should be formatting this?

Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/sharing/ShareActivity.kt]: that's right @parneet-guraya , not i see it correct 👍

**Comment by jesmrec** [comment]
Now, everything is fixed. @Aitorbp please approve if you think this is OK as well and we can merge it

**Review by Aitorbp — APPROVED** [review]
LGTM 🥇  Good job!
